streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStore.java

```diff
@@ -263,6 +266,78 @@ public VersionedRecord<byte[]> get(final Bytes key, final long asOfTimestamp) {
         return null;
     }
 
+    public VersionedRecordIterator<byte[]> get(final Bytes key, final long fromTimestamp, final long toTimestamp, final boolean isAscending) {
+
+        Objects.requireNonNull(key, "key cannot be null");
+        validateStoreOpen();
+
+        final List<VersionedRecord<byte[]>> queryResults = new ArrayList<>();
+
+        if (toTimestamp < observedStreamTime - historyRetention) {
+            // history retention exceeded. we still check the latest value store in case the
+            // latest record version satisfies the timestamp bound, in which case it should
+            // still be returned (i.e., the latest record version per key never expires).
+            final byte[] rawLatestValueAndTimestamp = latestValueStore.get(key);
+            if (rawLatestValueAndTimestamp != null) {
+                final long recordTimestamp = LatestValueFormatter.getTimestamp(rawLatestValueAndTimestamp);
+                if (recordTimestamp <= toTimestamp) {
+                    // latest value satisfies timestamp bound
+                    queryResults.add(new VersionedRecord<>(LatestValueFormatter.getValue(rawLatestValueAndTimestamp), recordTimestamp));
+                }
+            }
+
+            // history retention has elapsed and the latest record version (if present) does
+            // not satisfy the timestamp bound. return null for predictability, even if data
+            // is still present in segments.
+            if (queryResults.size() == 0) {
+                LOG.warn("Returning null for expired get.");
+            }
+            return new VersionedRecordIteratorImpl<>(queryResults.listIterator());
+        } else {
+            // take a RocksDB snapshot to return the segments content at the query time (in order to guarantee consistency)
+            final Snapshot snapshot = latestValueStore.getSnapshot();
```

Comment:
I think we need to get both snapshots for latestValeuStore and segmentStore at the same time ("atomically" protected by a lock) to avoid a race between both store.